src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1923,25 +1924,35 @@ private void PerformSchedulingActions(IEnumerable<ScheduleResponse> responses)
+                        var newNodes = new ConcurrentBag<NodeInfo>();
+                        GetNodeConfiguration();
+                        try
                         {
-                            NodeInfo createdNode = _nodeManager.CreateNode(GetNodeConfiguration(), response.RequiredNodeType);
-
-                            if (null != createdNode)
+                            Parallel.For(0, response.NumberOfNodesToCreate, i =>
                             {
-                                _noNodesActiveEvent.Reset();
-                                _activeNodes.Add(createdNode.NodeId);
-                                newNodes.Add(createdNode);
-                                ErrorUtilities.VerifyThrow(_activeNodes.Count != 0, "Still 0 nodes after asking for a new node.  Build cannot proceed.");
-                            }
-                            else
-                            {
-                                BuildEventContext buildEventContext = new BuildEventContext(0, Scheduler.VirtualNode, BuildEventContext.InvalidProjectInstanceId, BuildEventContext.InvalidProjectContextId, BuildEventContext.InvalidTargetId, BuildEventContext.InvalidTaskId);
-                                ((IBuildComponentHost)this).LoggingService.LogError(buildEventContext, new BuildEventFileInfo(String.Empty), "UnableToCreateNode", response.RequiredNodeType.ToString("G"));
+                                NodeInfo createdNode = _nodeManager.CreateNode(GetNodeConfiguration(), response.RequiredNodeType);
+
+                                if (createdNode != null)
+                                {
+                                    _noNodesActiveEvent.Reset();
+                                    _activeNodes.TryAdd(createdNode.NodeId, 0);
+                                    newNodes.Add(createdNode);
+                                    ErrorUtilities.VerifyThrow(_activeNodes.Count != 0, "Still 0 nodes after asking for a new node.  Build cannot proceed.");
+                                }
+                                else
+                                {
+                                    BuildEventContext buildEventContext = new BuildEventContext(0, Scheduler.VirtualNode, BuildEventContext.InvalidProjectInstanceId, BuildEventContext.InvalidProjectContextId, BuildEventContext.InvalidTargetId, BuildEventContext.InvalidTaskId);
+                                    ((IBuildComponentHost)this).LoggingService.LogError(buildEventContext, new BuildEventFileInfo(String.Empty), "UnableToCreateNode", response.RequiredNodeType.ToString("G"));
 
-                                throw new BuildAbortedException(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("UnableToCreateNode", response.RequiredNodeType.ToString("G")));
+                                    throw new BuildAbortedException(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("UnableToCreateNode", response.RequiredNodeType.ToString("G")));
```

Comment:
nit: AFAIK Parallel.For bundles all exceptions from all threads in an AggregateException. Might want to catch that and rethrow the containing exceptions to avoid changing the exceptions that API users will get.

Reply:
I'm confused. Is that not what I did? I have a try block around the parallel for, and the catch (of an AggregateException) just throws all the inner exceptions.

Reply:
Ah, my bad, didn't notice the loop in the catch block :blush: